Thread names in logs and deadlock debug tools (take 2)

jamesob opened the pull request:
Replaces #13099.

The goal here is the same as the previous PR, but this approach only includes threadnames in logs when we can make use of `thread_local`. We specifically disable `HAVE_THREAD_LOCAL` for platforms where its presence or implementation is questionable (Darwin, mingw32).

Instead of using a mutex-protected map to generate numeric suffixes (e.g. `httpworker.2`) we explicitly pass in suffixes during thread creation.

Under this changeset, logs look like this:
```
2018-06-13T21:43:28Z [bitcoin-qt-init] init message: Starting network threads...
2018-06-13T21:43:28Z [net] net thread start
2018-06-13T21:43:28Z [bitcoin-qt-init] init message: Done loading
2018-06-13T21:43:28Z [opencon] opencon thread start
2018-06-13T21:43:28Z [msghand] msghand thread start
2018-06-13T21:43:28Z [dnsseed] dnsseed thread start
2018-06-13T21:43:28Z [dnsseed] Loading addresses from DNS seeds (could take a while)
2018-06-13T21:43:28Z [dnsseed] 0 addresses found from DNS seeds
2018-06-13T21:43:28Z [dnsseed] dnsseed thread exit
2018-06-13T21:43:28Z [addcon] addcon thread start
2018-06-13T21:43:28Z [bitcoin-qt] GUI: Platform customization: "other"
2018-06-13T21:43:28Z [bitcoin-qt] GUI: PaymentServer::LoadRootCAs: Loaded  146  root certificates
2018-06-13T21:43:28Z [loadblk] Loaded 1 blocks from external file in 147ms
2018-06-13T21:43:28Z [loadblk] Reindexing finished
2018-06-13T21:43:28Z [loadblk] Imported mempool transactions from disk: 0 succeeded, 0 failed, 0 expired, 0 already there

```

sipa commented on src/threadval.cpp:
```diff
@@ -0,0 +1,146 @@
+    std::string m_name{"unnamed"};
+};
+
+#if defined(HAVE_THREAD_LOCAL)
```
Do we want to use thread local even on mingw (where you had problems earlier?)

theuni replied:
https://github.com/bitcoin/bitcoin/pull/13168/files#diff-defae1c29982e5afaf477729b6c7d9b4R11 :)

sipa commented on src/threadval.cpp:
```diff
@@ -0,0 +1,146 @@
+
+std::string thread_data::get_internal_name()
+{
+    void* data = pthread_getspecific(g_key);
```
(in this function and the next) `g_key` is uninitialized if called before `set_internal_name` is invoked. Perhaps you need to add the `pthread_once` call here as well?

If not, document that it's illegal to invoke these functions without knowing if `set_internal_name` has been called.

theuni replied:
Yep, good catch.

sipa commented on src/threadval.cpp:
```diff
@@ -0,0 +1,146 @@
+        static_cast<thread_data_type*>(data)->m_name = std::move(name);
+        return true;
+    }
+    data = new thread_data_type{internal_id++, std::move(name)};
```
The `internal_id` has different semantics here than in the thread_local case (here it's only incremented the first time a thread is given a name; in the above case it's incremented every time the name changes). Is that an issue?

theuni replied:
Yep. The thread_local impl is incorrect. We only want to increment if unset, so thread id's don't bounce around across renames.

sipa reviewed (commented):
A few nits/comments on the first commit. Also:
* Why the rename from threadval to threadutil? Ideally introduce the file with the preferred name in the first commit, or at least do the rename in a separate commit from other changes.
* Naming convention (functions should be CamelCase).
* get_internal_id seems entirely unused.

jamesob commented:
I figure "threadutil" might be a better name going forward because I anticipate as we continue to split stuff out of `util`, this unit might be a good home for things like `TraceThread`. If you or @theuni thinks we should stick with "threadval" I'm happy to revert, otherwise I'll split out the rename into a separate commit.

sipa commented:
@jamesob I have no opinion about the name. I'm only complaining about a rename in the middle of a PR in a commit that also introduces other changes.

theuni commented on src/threadutil.cpp:
```diff
@@ -0,0 +1,237 @@
+{
+    auto name = g_thread_data.m_name;
+
+    if (g_thread_data.m_name.empty()) {
```
Why the fallback?

jamesob replied:
There were lots of `"<unnamed>"` occurrences in the logs and so I figured it made sense to fall back to the process name if we hadn't explicitly named the thread (e.g. `bitcoind`, `bitcoin-qt`). Do you think it makes more sense to just explicitly call `rename` on those main threads?

jamesob replied:
I can see concerns with `get_internal_name` being a misnomer. Maybe we should introduce a `GetName` that first tries the internal name and then falls back to the process name?

theuni commented:
@jamesob You can just take the threadval files, don't worry about the commit attribution.

Agree with @sipa about the name change, it complicates the review a pretty good bit. Also no preference on naming. I originally named it "threadval" with the intention of making it a generic low-level interface for thread_local, but that's not worth the trouble.

jamesob committed:
tests: add threadutil tests

jamesob committed:
threads: prefix log messages with thread names

Introduce a new flag (`-logthreadnames`) which allows toggling
of this behavior.

jamesob committed:
threads: add thread names to deadlock debugging message

Also refactor CLockLocation to use an initialization list.

jamesob commented:
@sipa @theuni Thanks for the quick look and comments. I've pushed a reworked version of this branch which
- rewrites the commit history to be more easily reviewable (I hope),
- fixes naming conventions,
- fixes the uninitialized use of `g_key`,
- ensures each implementation of the thread_local behavior only bumps the internal ID accumulator on the first `SetInternalName` per thread,
- removes the unused `GetInternalID` function,
- removes fallback inference from process name when thread name is unset, and explicitly names main threads (`bitcoind`, `bitcoin-qt`), and
- fixes `/* Continued */` logline behavior by introducing *Continued macros which don't prefix log message with thread name.

TheBlueMatt commented:
What platforms still don't have thread_local? Can we just ifdef everything and upgrade our release build OSX SDK and get it in every release platform?

On May 4, 2018 9:01:28 PM UTC, jamesob <notifications@github.com> wrote:
>Replaces #13099.
>
>The goal here is the same as the previous PR, but this approach uses
>@theuni's gracefully degrading thread-local caching strategy to avoid
>(when possible) map lookup and mutex overhead during thread name
>lookup.
>
>Instead of using a mutex-protected map to generate numeric suffixes
>(e.g. `httpworker.2`) we explicitly pass in suffixes during thread
>creation.
>
>The logging format is also slightly different:
>```
>[bitcoind     ] 2018-05-04T20:58:13Z nBestHeight = 1296342
>[torcontrol   ] 2018-05-04T20:58:13Z torcontrol thread start
>[bitcoind     ] 2018-05-04T20:58:13Z Bound to [::]:18333
>[bitcoind     ] 2018-05-04T20:58:13Z Bound to 0.0.0.0:18333
>[bitcoind     ] 2018-05-04T20:58:13Z init message: Loading P2P
>addresses...
>[loadblk      ] 2018-05-04T20:58:13Z Leaving InitialBlockDownload
>(latching to false)
>[loadblk      ] 2018-05-04T20:58:13Z Imported mempool transactions from
>disk: 3 succeeded, 0 failed, 0 expired, 0 already there
>[bitcoind     ] 2018-05-04T20:58:13Z Loaded 19762 addresses from
>peers.dat  102ms
>[bitcoind     ] 2018-05-04T20:58:13Z init message: Loading banlist...
>[bitcoind     ] 2018-05-04T20:58:13Z init message: Starting network
>threads...
>[net          ] 2018-05-04T20:58:13Z net thread start
>[addcon       ] 2018-05-04T20:58:13Z addcon thread start
>[dnsseed      ] 2018-05-04T20:58:13Z dnsseed thread start
>[msghand      ] 2018-05-04T20:58:13Z msghand thread start
>[bitcoind     ] 2018-05-04T20:58:13Z init message: Done loading
>```
>You can view, comment on, or merge this pull request online at:
>
>  https://github.com/bitcoin/bitcoin/pull/13168
>
>-- Commit Summary --
>
>  * threads: add threadval files
>  * threads: hook up threadval for logging
>  * threads: add numeric suffixes to threads with the same name
>  * tests: add threadutil naming tests
>  * threads: add thread names to deadlock debugging tools
>
>-- File Changes --
>
>    M src/Makefile.am (2)
>    M src/Makefile.test.include (1)
>    M src/httpserver.cpp (11)
>    M src/init.cpp (7)
>    M src/logging.cpp (11)
>    M src/sync.cpp (14)
>    M src/test/test_bitcoin.cpp (2)
>    A src/test/threadutil_tests.cpp (90)
>    A src/threadutil.cpp (237)
>    A src/threadutil.h (62)
>    M src/util.cpp (21)
>    M src/util.h (6)
>    M src/validation.cpp (5)
>    M src/validation.h (8)
>
>-- Patch Links --
>
>https://github.com/bitcoin/bitcoin/pull/13168.patch
>https://github.com/bitcoin/bitcoin/pull/13168.diff
>
>-- 
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/bitcoin/bitcoin/pull/13168

sipa commented:
General comment: never @mention someone directly in a commit. They'll be spammed by every fork on github when those commits are merged by them.

sipa commented on src/threadutil.cpp:
```diff
@@ -0,0 +1,213 @@
+void thread_util::Rename(std::string name)
+{
+    SetProcessName(name.c_str());
+    SetInternalName(name);
```
This can be `SetInternalName(std::move(name))`.

sipa commented on src/threadutil.cpp:
```diff
@@ -0,0 +1,213 @@
+bool thread_util::SetInternalName(std::string name)
+{
+    static std::atomic<long> id_accumulator{0};
+    static thread_local thread_id{id_accumulator++};
```
This increments id every time the name is set (regardless of whether it had a name already).

jamesob replied:
Doesn't the `static thread_local` attribute prevent that?

sipa replied:
Oh, yes indeed. Ignore me

sipa commented on src/threadutil.cpp:
```diff
@@ -0,0 +1,213 @@
+static bool EnsureKeyCreated()
+{
+    static pthread_once_t key_once = PTHREAD_ONCE_INIT;
+    return pthread_once(&key_once, make_key) ? false : true;
```
Can be written as `!pthread_once(&key_once, make_key);`.

sipa commented on src/sync.cpp:
```diff
@@ -35,23 +39,27 @@ void PrintLockContention(const char* pszName, const char* pszFile, int nLine)
         sourceFile = pszFile;
         sourceLine = nLine;
         fTry = fTryIn;
+        thread_name = thread_name_;
```
This can be written as `thread_name = std::move(thread_name_)` to avoid a copy.

sipa reviewed (commented):
Some more comments.

Overall: why do we need `LogPrint*Continued`? It should be implied by a lack of newline on the previous log statement (which is already used to determine whether to log the time).

jamesob commented:
We need `LogPrint*Continued` to prevent prefixing a line continuation with a thread name. Thanks for the look, will incorporate the feedback tomorrow.

jamesob commented:
@sipa ah, sorry: I missed your point about the LogPrint macros. I see now that we can just move thread name attachment to reuse the existing `m_started_new_line` stuff, thanks. I've removed the last commit and worked in your other feedback.

[`jamesob/threadnames2.3 -> jamesob/threadnames2.4`](https://gist.github.com/jamesob/2aebc372f6421e6888ff2e5c29ac71b0)

jamesob commented:
@TheBlueMatt honestly I'm not sure about the thread_local stuff; all I know is that mingw32 wasn't behaving well with its use (as noted [in the comment](https://github.com/bitcoin/bitcoin/pull/13168/files#diff-defae1c29982e5afaf477729b6c7d9b4R13)) - @theuni is the authority there.

sipa commented on src/threadutil.cpp:
```diff
@@ -0,0 +1,213 @@
+ *
+ *   https://gist.github.com/jamesob/fe9a872051a88b2025b1aa37bfa98605
+ */
+#undef HAVE_THREAD_LOCAL
```
I think this line may be unintentional.

jamesob replied:
I don't think so - see the comment above; sounds like @theuni wants to do more testing with dynamic linking before we use `thread_local` outside the context of debug code.

sipa replied:
Unsure if we should have the code in in that case, as it's untestable (it also doesn't compile currently).

sipa commented on src/threadutil.cpp:
```diff
@@ -0,0 +1,213 @@
+#endif
+
+    char threadname_buff[16];
+    char* pthreadname_buff = (char*)(&threadname_buff);
```
I don't think you need this variable; using `threadname_buff` will automatically decay to `(char *)`.

jamesob commented:
Changed with more feedback from @sipa. 

[`jamesob/threadnames2.4` -> `.5`](https://gist.github.com/jamesob/5604eb4586f8eb3401659566e17632f8)

theuni commented:
@TheBlueMatt unfortunately it's not the SDK, it's Apple's clang itself that had thread_local patched out until xcode 9.0, released in June 2017. And to make it worse, they held off on supporting it because they "didn't want to commit to an inefficient abi". It's that reasoning that makes me believe that a build with an upstream clang may have compatibility issues with end-users' libc/libstdc++. A few tests might shed some light, but even then, I'd worry about one-off version issues.

So, options are:
- Move forward with thread_local, require mac builders to upgrade to xcode 9, hope for no abi issues
- Wait for xcode 9 to be a reasonable minimum, same as above
- \#ifndef \_\_APPLE\_\_ ...
- No thread_local for now

Those options ignore mingw, with which @jamesob has also reported some thread_local issues, but I haven't investigated those.

@sipa I left the thread_local code but undef'd the variable in case we want to go for \#3. Otherwise, I begrudgingly agree that it probably shouldn't be present until it's usable.

sipa commented on src/threadutil.cpp:
```diff
@@ -0,0 +1,212 @@
+bool thread_util::SetInternalName(std::string name)
+{
+    static std::atomic<long> id_accumulator{0};
+    static thread_local thread_id{id_accumulator++};
```
This declaration misses a type (when you uncomment the `#undef HAVE_THREAD_LOCAL`).

jamesob replied:
More evidence that we should remove this code until we need it. I'll strip it out.

jamesob commented:
I've pushed changes removing the unused `thread_local` implementation.

[`jamesob/threadnames.5 -> .6`](https://gist.github.com/jamesob/7759579c2ece06e9fa7db43652a47600)

MarcoFalke commented:
Needs rebase due to mege of #13236

ryanofsky commented:
> @TheBlueMatt and @ryanofsky gave feedback offline that we should only include a (very simple) thread_local implementation for this change

I'm happy with either approach. I just wanted to suggest that not logging thread names on platforms that don't support thread_local could be ok and simplify things.

jamesob commented:
I've pushed a simplified rework of this branch which removes all non-`thread_local` implementations of the threadutil stuff; we just won't support threadnames if the platform in question doesn't support thread_local. I've also added a flag to disable threadname prefixes (`-logthreadnames`) a la `-logtimestamps`, and have added related release notes.

This should be ready for review. One outstanding question I have is where we should be doing the conditional `#undef HAVE_THREAD_LOCAL` -- `threadutil.cpp` doesn't seem like the right place.

jamesob committed:
disable thread_local on unreliable platforms

See discussions here:

- https://github.com/bitcoin/bitcoin/pull/11722#pullrequestreview-79322658
- https://github.com/bitcoin/bitcoin/pull/13168#issuecomment-387181155

skeees commented:
concept ACK (I'm already finding this useful) - a quick note that this breaks combine logs.py (but trivial fix - just need to update the regexp that identifies timestamps)

MarcoFalke commented:
Needs rebase due to merge of #13341

jamesob commented:
Thanks for the heads-up, @MarcoFalke. I've rebased and pushed a fix to combine_logs.py (thanks @skeees). This is ready for review-for-merge.

jnewbery commented on src/logging.cpp:
```diff
@@ -190,29 +197,37 @@ std::string BCLog::Logger::LogTimestampStr(const std::string &str)
 void BCLog::Logger::LogPrintStr(const std::string &str)
 {
-    std::string strTimestamped = LogTimestampStr(str);
+    std::string strPrefixed = LogThreadnameStr(LogTimestampStr(str));
```
It seems more natural to me to have the thread name after the timestamp.

This way round breaks combine_logs without a fix. I imagine other log parsers expect the timestamp at the start of the line as well.

jamesob replied:
Fixed

jnewbery commented on src/threadutil.h:
```diff
@@ -0,0 +1,57 @@
+    void Rename(std::string);
+
+    /**
+     * Set the thread's internal (in-memory) name; used e.g. for identification in
```
Comment is incorrect. Should be `Get the thread's ...`, or `@return the thread's...`

jamesob replied:
Fixed.

jnewbery commented on test/functional/combine_logs.py:
```diff
@@ -12,8 +12,8 @@
-# Matches on the date format at the start of the log event
-TIMESTAMP_PATTERN = re.compile(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{6}Z")
+# Matches on the date format or threadname prefix at the start of the log event
+PREFIX_PATTERN = re.compile(r"^(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{6}Z|\[\w+\s*\] )")
```
The pattern doesn't work for the `[httpworker.n ]` thread logs:

```
/tmp/user/1000/testb2liucdnnode0 [http         ] 2018-06-07T17:33:43.612724Z Received a POST request for / from 127.0.0.1:34830

[httpworker.2 ] 2018-06-07T17:33:43.612865Z ThreadRPCServer method=getblockcount user=__cookie__ 
 node0 [http         ] 2018-06-07T17:33:43.613814Z Received a POST request for / from 127.0.0.1:34830

[httpworker.0 ] 2018-06-07T17:33:43.613905Z ThreadRPCServer method=listwallets user=__cookie__ 
 node0 [http         ] 2018-06-07T17:33:43.614514Z Received a POST request for / from 127.0.0.1:34830
```

This problem goes away if you leave the timestamp at the start of the log string.

jnewbery reviewed (commented):
Generally looks good. A couple of comments inline.

the txindex thread is not named. I think you'll need to call `thread_util::Rename()` in `TxIndex::Start()` (in txindex.cpp).

jamesob commented on src/util.cpp:
```diff
@@ -9,6 +9,7 @@
 #include <random.h>
 #include <serialize.h>
 #include <utilstrencodings.h>
+#include <threadutil.h>
```
I think so: `thread_util::Rename` is used on [line 318](https://github.com/jamesob/bitcoin/blob/e6454374033760d93a33619bea1451573718dd3d/src/util.h#L318).

jamesob replied:
Oh no, you're right: I confused this with util.h. Thanks!

jamesob commented on src/util.h:
```diff
@@ -309,15 +310,12 @@ std::string HelpMessageOpt(const std::string& option, const std::string& message
 /**
  * .. and a wrapper that just calls func once
  */
 template <typename Callable> void TraceThread(const char* name,  Callable func)
```
Only problem is that'd create a circular dep with `logging` and `threadutil` due to use of `LogPrintf` here. I agree that in concept that's where it should live, but it doesn't look trivial to avoid that circular dep.

ryanofsky commented on src/qt/bitcoin.cpp:
```diff
@@ -282,6 +283,7 @@ void BitcoinCore::initialize()
     try
     {
         qDebug() << __func__ << ": Running initialization in thread";
+        thread_util::Rename("bitcoin-qt");
```
In commit "threads: explicitly name main threads" (56deb18c8bc36de5bb49c7b1db27adfa80ce7169)

This is giving 2 different threads (the `main()` thread, and the initialization thread spawned in `BitcoinApplication::startThread`) the same name. It might be better to distinguish this thread by calling it `qt-init` or something instead of `bitcoin-qt`.

This is also kind of an awkward place to be name the thread, since it isn't in a standard thread entry point or close to where the thread is created. Maybe it would make sense for us to have a custom QThread subclass that supports named threads, and calls `thread_util::Rename` in [`exec()`](http://doc.qt.io/qt-5/qthread.html#exec). Or maybe we could have write a little helper object that attaches to a `QThread`, listens for the [`started`](http://doc.qt.io/qt-5/qthread.html#started) signal, and sets the thread name in the signal handler.

jamesob replied:
Changed the name of this thread to `bitcoin-qt-init`.

Agree with you that this is an awkward spot for naming. Do you think your proposed fix should be done along with this changeset or maybe as a follow-up PR?

ryanofsky replied:
https://github.com/bitcoin/bitcoin/pull/13168#discussion_r195165776

> Do you think your proposed fix should be done along with this changeset or maybe as a follow-up PR?

I might opt to do it in a different PR because QThread is also used in rpcconsole.cpp and bitcoin.cpp, and it might make sense to update those places to set thread names in the same way. But I don't have a real opinion. It seems fine to make the change in this PR, in another PR, or not at all.

ryanofsky reviewed (commented):
utACK e6454374033760d93a33619bea1451573718dd3d

jamesob committed:
threads: introduce threadutil, refactor thread naming

This work is prerequisite to attaching thread names to log lines and deadlock
debug utilities. threadutil allows setting of an "internal" threadname per
thread on platforms where thread_local is available.

This commit also moves RenameThread() out of util and adds a numeric suffix to
disambiguate between threads with the same name. It explicitly names a few
main threads using the new Rename() utility.

jamesob commented on configure.ac:
```diff
@@ -759,8 +759,21 @@ AC_LINK_IFELSE([AC_LANG_SOURCE([
+        AC_MSG_RESULT(no)
+        ;;
+      *darwin*)
+        AC_MSG_RESULT(no)
```
Added, thanks.

jamesob commented on src/threadutil.cpp:
```diff
@@ -0,0 +1,99 @@
+#include <sys/prctl.h> // For prctl, PR_SET_NAME, PR_GET_NAME
+#endif
+
+void thread_util::SetProcessName(const char* name)
```
Fixed

jamesob commented on src/httpserver.cpp:
```diff
@@ -281,7 +282,7 @@ static void http_reject_request_cb(struct evhttp_request* req, void*)
 /** Event dispatcher thread */
 static bool ThreadHTTP(struct event_base* base, struct evhttp* http)
 {
-    RenameThread("bitcoin-http");
```
Took your advice; am now prefixing threadname with `"bitcoin-"` for `SetThreadName` calls.

jamesob commented on src/threadutil.cpp:
```diff
@@ -0,0 +1,99 @@
+ */
+#if defined(HAVE_THREAD_LOCAL)
+
+struct thread_data_type
```
`thread_data_type` has been removed entirely.

jamesob commented on src/init.cpp:
```diff
@@ -457,6 +457,7 @@ void SetupServerArgs()
     gArgs.AddArg("-help-debug", "Show all debugging options (usage: --help -help-debug)", false, OptionsCategory::DEBUG_TEST);
     gArgs.AddArg("-logips", strprintf("Include IP addresses in debug output (default: %u)", DEFAULT_LOGIPS), false, OptionsCategory::DEBUG_TEST);
     gArgs.AddArg("-logtimestamps", strprintf("Prepend debug output with timestamp (default: %u)", DEFAULT_LOGTIMESTAMPS), false, OptionsCategory::DEBUG_TEST);
+    gArgs.AddArg("-logthreadnames", strprintf("Prepend debug output with name of the originating thread (default: %u)", DEFAULT_LOGTHREADNAMES), false, OptionsCategory::DEBUG_TEST);
```
Fixed

jamesob commented on src/logging.cpp:
```diff
@@ -190,29 +197,37 @@ std::string BCLog::Logger::LogTimestampStr(const std::string &str)
     return strStamped;
 }
 
+std::string BCLog::Logger::LogThreadnameStr(const std::string &str)
```
Now that we've reversed the order of threadname and log timestamp (threadname follows timestamp), I've only modified `LogTimestampStr` per your suggestion since it's the only function of the two to receive an rvalue as input. `LogThreadnameStr` receives an lvalue from `LogPrintStr` which itself receives an lvalue from `logging.h:LogPrintf`; I may be mistaken here, but the optimization you describe wouldn't apply in this case.

jamesob commented on src/threadutil.cpp:
```diff
@@ -0,0 +1,99 @@
+
+struct thread_data_type
+{
+    constexpr thread_data_type() = default;
```
Fixed; thread_data_type has been removed.

jamesob commented on src/threadutil.cpp:
```diff
@@ -0,0 +1,99 @@
+bool thread_util::SetInternalName(std::string name)
+{
+    static std::atomic<long> id_accumulator{0};
+    static thread_local long thread_id{id_accumulator++};
```
Yep, dropped. The changeset has been substantially simplified to use a single `thread_local std::string g_thread_name` variable instead of a struct.

jamesob commented on src/logging.cpp:
```diff
@@ -190,29 +197,37 @@ std::string BCLog::Logger::LogTimestampStr(const std::string &str)
+    }
+    std::string thread_name = thread_util::GetInternalName();
+    // The longest thread name (with numeric suffix) we have at the moment is 13 characters.
+    thread_name.resize(13, ' ');
```
We're now no longer resizing the internal threadname; thanks.

jamesob commented on src/sync.cpp:
```diff
@@ -37,23 +39,27 @@ void PrintLockContention(const char* pszName, const char* pszFile, int nLine)
         sourceFile = pszFile;
         sourceLine = nLine;
         fTry = fTryIn;
+        thread_name = std::move(thread_name_);
```
Took @ryanofsky's advice and refactored this constructor to use an initialization list.

jamesob commented on src/httpserver.cpp:
```diff
@@ -328,9 +329,9 @@ static bool HTTPBindAddresses(struct evhttp* http)
+static void HTTPWorkQueueRun(WorkQueue<HTTPClosure>* queue, int worker_num)
 {
-    RenameThread("bitcoin-httpworker");
+    thread_util::Rename(("httpworker." + std::to_string(worker_num)).c_str());
```
This commit seems tough to split up in a meaningful way. Hopefully the reduced size of the `threadutil` change (now that we aren't doing anything with a generic thread data struct) makes it easier to review, but it'll take a lot of rework to stage the changes out as you suggest. I'm happy to do this if you think it's worthwhile, but wanted to check before doing so to see if the simplifications make a difference for this recommendation.

jamesob commented on src/test/threadutil_tests.cpp:
```diff
@@ -0,0 +1,94 @@
+
+BOOST_FIXTURE_TEST_SUITE(threadutil_tests, BasicTestingSetup)
+
+BOOST_AUTO_TEST_CASE(threadutil_test_process_rename_serial)
```
Removed this test.

jamesob commented on doc/release-notes-pr13168.md:
```diff
@@ -0,0 +1,5 @@
+Thread names in logs
+--------------------
+
+Log lines are now prefixed by default with the name of the thread that caused
```
Fixed, thanks.

jamesob commented:
Thanks very much for the reviews, @jnewbery @ryanofsky. I've rebased and pushed changes incorporating your good feedback. The changeset has been simplified; instead of the generic thread data struct (which was more or less a vestige from when we were trying to provide threadnames for all platforms, even those not supporting `thread_local`), we're now just using a single thread_local variable for the threadname. I've also moved the threadname back to following timestamp as suggested.

jamesob commented:
Looks like I botched the rebase; https://github.com/bitcoin/bitcoin/pull/13168/commits/d930487b3dc398f3ac93c00324fff027a4c61588 shouldn't be in this PR... Fixing.

jamesob commented:
Botched rebase fixed.

ryanofsky commented on configure.ac:
```diff
@@ -763,8 +763,23 @@ AC_LINK_IFELSE([AC_LANG_SOURCE([
+        AC_MSG_RESULT(no)
+        ;;
+      *darwin*)
+        # TODO enable therad_local on later versions of Darwin where it is
```
In commit "disable thread_local on unreliable platforms" (696bdb751055c01c645dc95df486d56a873325de)

Little spelling mistake

jamesob replied:
D'oh. Fixed.

ryanofsky commented on src/bitcoind.cpp:
```diff
@@ -57,6 +58,8 @@ static bool AppInit(int argc, char* argv[])
 {
     bool fRet = false;
 
+    thread_util::Rename("bitcoind");
```
In commit "threads: explicitly name main threads" (cf82eda8aec147a0d24ade9263f8df0708ff02bb)

Since OS thread names are now prefixed with "bitcoin-", I guess these will show up as "bitcoin-bitcoind", "bitcoin-bitcoin-qt-init", and ""bitcoin-bitcoin-qt" in top? This seems ok, but a little redundant. You could consider just calling `Rename()` with "main" and "init" so the OS names would be "bitcoin-main" and "bitcoin-init".

jamesob replied:
Done, thanks. Non-internal thread names now look like 
```sh
 $ pstree 12441

bitcoin-init─┬─{bitcoin-addcon}
             ├─{bitcoin-dnsseed}
             ├─4*[{bitcoin-httpwor}]
             ├─{bitcoin-http}
             ├─{bitcoin-init}
             ├─{bitcoin-loadblk}
             ├─{bitcoin-msghand}
             ├─{bitcoin-net}
             ├─{bitcoin-opencon}
             ├─{bitcoin-schedul}
             ├─7*[{bitcoin-scriptc}]
             └─{bitcoin-torcont}
```

ryanofsky reviewed (commented):
utACK cf82eda8aec147a0d24ade9263f8df0708ff02bb. Lots of little changes since the last review, all discussed above. Main user visible changes are putting thread names after timestamps on log lines and not padding log thread names with spaces.

jamesob commented:
Thanks for the continued review, @ryanofsky. Pushed a few small changes addressing your feedback.

```sh
 $ git diff threadnames2.12..threadnames2.11 | cat

diff --git a/configure.ac b/configure.ac
index 6873326..31c8c66 100644
--- a/configure.ac
+++ b/configure.ac
@@ -771,7 +771,7 @@ AC_LINK_IFELSE([AC_LANG_SOURCE([
         AC_MSG_RESULT(no)
         ;;
       *darwin*)
-        # TODO enable thread_local on later versions of Darwin where it is
+        # TODO enable therad_local on later versions of Darwin where it is
         # supported (per https://stackoverflow.com/a/29929949)
         AC_MSG_RESULT(no)
         ;;
diff --git a/src/bitcoind.cpp b/src/bitcoind.cpp
index ee99099..066e83d 100644
--- a/src/bitcoind.cpp
+++ b/src/bitcoind.cpp
@@ -58,7 +58,7 @@ static bool AppInit(int argc, char* argv[])
 {
     bool fRet = false;

-    thread_util::Rename("init");
+    thread_util::Rename("bitcoind");

     //
     // Parameters
```

jamesob commented:
Oops -- missed the QT `Rename` calls in the last change. Pushed: 
```sh
 $ git diff threadnames2.12..threadnames2.13 | cat

diff --git a/src/qt/bitcoin.cpp b/src/qt/bitcoin.cpp
index d3f97f7..f13e655 100644
--- a/src/qt/bitcoin.cpp
+++ b/src/qt/bitcoin.cpp
@@ -283,7 +283,7 @@ void BitcoinCore::initialize()
     try
     {
         qDebug() << __func__ << ": Running initialization in thread";
-        thread_util::Rename("bitcoin-qt-init");
+        thread_util::Rename("qt-init");
         bool rv = m_node.appInitMain();
         Q_EMIT initializeResult(rv);
     } catch (const std::exception& e) {
@@ -568,7 +568,7 @@ static void SetupUIArgs()
 int main(int argc, char *argv[])
 {
     SetupEnvironment();
-    thread_util::Rename("bitcoin-qt");
+    thread_util::Rename("main");

     std::unique_ptr<interfaces::Node> node = interfaces::MakeNode();
```

jnewbery commented on src/logging.cpp:
```diff
@@ -168,7 +174,7 @@ std::vector<CLogCategoryActive> ListActiveLogCategories()
 }
 
-std::string BCLog::Logger::LogTimestampStr(const std::string &str)
+std::string BCLog::Logger::LogTimestampStr(std::string str)
```
I don't think you need to change the signature here. Any reason to change from a const reference?

jamesob replied:
See https://github.com/bitcoin/bitcoin/pull/13168#discussion_r195141701

jnewbery reviewed (commented):
A few nits inline.

The txindex thread is not named. Perhaps you lost that when rebasing on #13243

I've tested and other than those nits, it looks good!

jamesob commented on src/threadutil.cpp:
```diff
@@ -0,0 +1,80 @@
+ * Set the in-memory internal name for this thread. Does not affect the process
+ * name.
+ */
+static bool SetInternalName(const std::string& name)
```
`g_thread_name` doesn't exist as a symbol in non-`thread_local` enabled environments, so basically our choice is one `#if defined(...)` or three inline. I figure one is preferable, but if you feel strongly I'm happy to change.

jnewbery replied:
No, I don't feel strongly. Both are fine!

jamesob commented:
Thanks for the look, @jnewbery. Pushed [`threadnames2.13` -> `threadnames2.14`](https://gist.github.com/jamesob/4b68362c5b47c6ce46ea712ebb9995e2). 

It turns out the txindex thread *is* actually named ([here](https://github.com/bitcoin/bitcoin/blob/master/src/index/base.cpp#L267-L268) - GetName() is passed into TraceThread(), which then calls Rename()):

```
2018-06-22T20:36:46Z [txindex] txindex thread start
2018-06-22T20:36:46Z [qt-init] init message: Loading wallet...
2018-06-22T20:36:46Z [txindex] txindex is enabled
2018-06-22T20:36:46Z [txindex] txindex thread exit
```

jnewbery commented:
Do you know what thread these leveldb logs come from:

```
2018-06-25T14:20:09Z [init] Bitcoin Core version v0.16.99.0-unk (release build)
2018-06-25T14:20:09Z [init] InitParameterInteraction: parameter interaction: -whitelistforcerelay=1 -> setting -whitelistrelay=1
2018-06-25T14:20:09Z [init] Assuming ancestors of block 0000000000000000005214481d2d96f898e3d5416e43359c145944a909d242e0 have valid signatures.
2018-06-25T14:20:09Z [init] Setting nMinimumChainWork=000000000000000000000000000000000000000000f91c579d57cad4bc5278cc
2018-06-25T14:20:09Z [init] Using the 'sse4(1way+4way)' SHA256 implementation
2018-06-25T14:20:09Z [init] Using RdRand as an additional entropy source
2018-06-25T14:20:09Z [init] Default data directory /home/ubuntu/.bitcoin
2018-06-25T14:20:09Z [init] Using data directory /home/ubuntu/.bitcoin
2018-06-25T14:20:09Z [init] Using config file /home/ubuntu/.bitcoin/bitcoin.conf
2018-06-25T14:20:09Z [init] Using at most 125 automatic connections (1024 file descriptors available)
2018-06-25T14:20:09Z [init] Using 16 MiB out of 32/2 requested for signature cache, able to store 524288 elements
2018-06-25T14:20:09Z [init] Using 16 MiB out of 32/2 requested for script execution cache, able to store 524288 elements
2018-06-25T14:20:09Z [init] Using 2 threads for script verification
2018-06-25T14:20:09Z [init] Allowing HTTP connections from: 127.0.0.0/8 ::1/128 127.0.0.1/32 
2018-06-25T14:20:09Z [init] Binding RPC on address :: port 44553
2018-06-25T14:20:09Z [init] Binding RPC on address 0.0.0.0 port 44553
2018-06-25T14:20:09Z [init] Binding RPC on address 0.0.0.0 port 44553 failed.
2018-06-25T14:20:09Z [init] Initialized HTTP server
2018-06-25T14:20:09Z [init] HTTP: creating work queue of depth 16
2018-06-25T14:20:09Z [init] Starting RPC
2018-06-25T14:20:09Z [init] Starting HTTP RPC server
2018-06-25T14:20:09Z [init] Config options rpcuser and rpcpassword will soon be deprecated. Locally-run instances may remove rpcuser to use cookie-based auth, or may be replaced with rpcauth. Please see share/rpcuser for rpcauth auth generation.
2018-06-25T14:20:09Z [init] Registering HTTP handler for / (exactmatch 1)
2018-06-25T14:20:09Z [init] Registering HTTP handler for /wallet/ (exactmatch 0)
2018-06-25T14:20:09Z [init] Starting HTTP server
2018-06-25T14:20:09Z [init] HTTP: starting 4 worker threads
2018-06-25T14:20:09Z [init] Using wallet directory /home/ubuntu/.bitcoin
2018-06-25T14:20:09Z [init] init message: Verifying wallet(s)...
2018-06-25T14:20:09Z [init] Using BerkeleyDB version Berkeley DB 4.8.30: (April  9, 2010)
2018-06-25T14:20:09Z [init] Using wallet wallet.dat
2018-06-25T14:20:09Z [init] BerkeleyEnvironment::Open: LogDir=/home/ubuntu/.bitcoin/database ErrorFile=/home/ubuntu/.bitcoin/db.log
2018-06-25T14:20:09Z [scheduler] scheduler thread start
2018-06-25T14:20:09Z [http] Entering http event loop
2018-06-25T14:20:09Z [init] net: setting try another outbound peer=false
2018-06-25T14:20:09Z [init] Cache configuration:
2018-06-25T14:20:09Z [init] * Using 2.0MiB for block index database
2018-06-25T14:20:09Z [init] * Using 56.0MiB for transaction index database
2018-06-25T14:20:09Z [init] * Using 8.0MiB for chain state database
2018-06-25T14:20:09Z [init] * Using 384.0MiB for in-memory UTXO set (plus up to 286.1MiB of unused mempool space)
2018-06-25T14:20:09Z [init] init message: Loading block index...
2018-06-25T14:20:09Z [init] LevelDB using max_open_files=1000 (default=1000)
2018-06-25T14:20:09Z [init] Opening LevelDB in /home/ubuntu/.bitcoin/blocks/index
2018-06-25T14:20:09Z [init] leveldb: Recovering log #247
2018-06-25T14:20:10Z [init] leveldb: Level-0 table #249: started
2018-06-25T14:20:10Z [init] leveldb: Level-0 table #249: 5789826 bytes OK
2018-06-25T14:20:10Z [init] leveldb: Recovering log #248
2018-06-25T14:20:10Z [init] leveldb: Level-0 table #250: started
2018-06-25T14:20:10Z [init] leveldb: Level-0 table #250: 193 bytes OK
2018-06-25T14:20:10Z [init] leveldb: Delete type=0 #247
2018-06-25T14:20:10Z [init] leveldb: Delete type=3 #244
2018-06-25T14:20:10Z [init] leveldb: Delete type=0 #248
2018-06-25T14:20:10Z [init] Opened LevelDB successfully
2018-06-25T14:20:10Z [init] Using obfuscation key for /home/ubuntu/.bitcoin/blocks/index: 0000000000000000
2018-06-25T14:20:10Z [] leveldb: Compacting 4@0 + 7@1 files
2018-06-25T14:20:10Z [] leveldb: Generated table #252@0: 15702 keys, 2141621 bytes
2018-06-25T14:20:10Z [] leveldb: Generated table #253@0: 8086 keys, 1117181 bytes
2018-06-25T14:20:10Z [] leveldb: Generated table #254@0: 15629 keys, 2139993 bytes
2018-06-25T14:20:10Z [] leveldb: Generated table #255@0: 6885 keys, 942686 bytes
2018-06-25T14:20:10Z [] leveldb: Generated table #256@0: 15662 keys, 2141826 bytes
2018-06-25T14:20:10Z [] leveldb: Generated table #257@0: 5521 keys, 755359 bytes
2018-06-25T14:20:10Z [] leveldb: Generated table #258@0: 5260 keys, 711244 bytes
2018-06-25T14:20:10Z [] leveldb: Compacted 4@0 + 7@1 files => 9949910 bytes
2018-06-25T14:20:10Z [] leveldb: compacted to: files[ 0 7 35 0 0 0 0 ]
```

jnewbery commented:
Tested ACK 18928a194fa351e07eb808eaea01792ceb1d2093, modulo the question above about the leveldb logs.

If there are any further changes to this PR, I'd suggest squashing commit *threads: explicitly name main threads* into commit *threads: introduce threadutil, refactor thread naming*.

jamesob commented:
@jnewbery Thanks for the follow-up. Turns out that the blank names contained in your logs come from a [compaction](https://github.com/jamesob/bitcoin/blob/18928a194fa351e07eb808eaea01792ceb1d2093/src/leveldb/db/db_impl.cc#L657-L660) thread [started from within the leveldb source tree](https://github.com/jamesob/bitcoin/blob/18928a194fa351e07eb808eaea01792ceb1d2093/src/leveldb/util/env_posix.cc#L622-L629), so I don't think there's much we can do there.

I'll be sure to squash the commits you reference pending more modifications.

ryanofsky reviewed (commented):
utACK 18928a194fa351e07eb808eaea01792ceb1d2093. Only minor changes since last review: updating init thread names, dropping bool return values, fixing spelling mistake.

jamesob commented:
Rebased

jamesob commented on src/logging.cpp:
```diff
@@ -5,6 +5,12 @@
 
 #include <logging.h>
 #include <utiltime.h>
+#include <util.h>
```
Good catch; this was a relic from previous code.

jamesob commented on src/validation.cpp:
```diff
@@ -1650,8 +1651,8 @@ static bool WriteUndoDataForBlock(const CBlockUndo& blockundo, CValidationState&
-void ThreadScriptCheck() {
-    RenameThread("bitcoin-scriptch");
+void ThreadScriptCheck(int worker_num) {
+    thread_util::Rename(("scriptch." + std::to_string(worker_num)).c_str());
```
Ah thanks, will fix the `httpworker` rename calls as well.

jamesob commented:
Thanks for the look, @sipa. I've incorporated your suggestions and squashed commits as suggested by @jnewbery earlier. 

Diff here: [threadnames2.15 -> .16](https://gist.github.com/jamesob/eb392fa435782d3b34cf04de90959ffa)

jnewbery commented:
Tested ACK 91aef702c90d53c8f8dd53dfacb5d5f3d0cb2b91

jamesob commented:
I benchmarked this branch (as rebased on top of master) with [bitcoinperf]() and it's worth noting that we see a roughly 1.4% slowdown over master ([details here](https://gist.github.com/jamesob/8115f1d24086580d4dbd54dba4b27047)) during IBD up to height 200,000.

ryanofsky reviewed (commented):
> we see a roughly 1.4% slowdown over master

What's the implication of this? Do you want to to try to optimize the change, or abandon it, or turn it off by default? Does the slowdown even happen in the default configuration, or only when extra debugging is enabled?

tACK 91aef702c90d53c8f8dd53dfacb5d5f3d0cb2b91. I just ran `bitcoind -regtest` and looked at ps thread names and log output. Only changes since my previous review were a rebase and tweaked thread names.

jamesob commented:
@ryanofsky thanks for the look. I think given the observed slowdown, maybe disabling by default is the way to go. I don't see any obvious means of improving performance, though I haven't taken to profiling tools.

fanquake commented:
Removing "Up for grabs" given #15849.
